### PR TITLE
Support more generalized macro for enum

### PR DESCRIPTION
**Pull request by SeokminHong**
Hi 👋 

While using Rustler, I've wanted to use more complex enum types for Elixir bindings. So I tried to implement it, which can be used like the following snippet.

```rust
#[derive(rustler::NifEnum)]
enum Test {
  Foo{ x: i32, y: i32 }, // Allow named enum
  Bar(String),
  Baz(String),  // Allow duplicated types
  Qux
}

#[rustler::nif]
fn claim_test(value: i32) -> Test {
  match value {
    0 => Test::Foo{ x: 1, y: 2 },        // %{__enum__: :foo, x: 1, y: 2}
    1 => Test::Bar("Hello".to_string()), // {:bar, "Hello"}
    2 => Test::Baz("world".to_string()), // {:baz, "world"}
    3 => Test::Qux,                      // :qux
    _ => panic!("")
  }
}
```

If there are more idiomatic ways to bind, please let me know!

Also, I didn't realize there was a related issue, #426. I'm really sorry about that 🙏

### CODE REVIEW

**Comment by evnu** [comment]
> ``` rust
> #[derive(rustler::NifStruct)]
> enum Test {
> ```

I think annotating an enum with `NifStruct` might be confusing.

Side-note: should we also deprecate `NifUnitEnum` if we go with this PR? I think `NifEnum` would also cover `NifUnitEnum`.

**Comment by turion** [comment]
> > ```rust
> > #[derive(rustler::NifStruct)]
> > enum Test {
> > ```
> 
> I think annotating an enum with `NifStruct` might be confusing.

I found it clear because I thought that `Struct` refers to the elixir struct. But yes, if you take it to refer to Rust structs, it doesn't make sense anymore.

**Comment by SeokminHong** [comment]
@turion Thank you for your reply. I also like your suggestion, but as @evnu said, the `Enum` idiom for Rust's enum would be more appealing to me.

---

> Side-note: should we also deprecate `NifUnitEnum` if we go with this PR? I think `NifEnum` would also cover `NifUnitEnum`.

@evnu If this PR is approved, I think it will.

---

I also want to have some conversation about encoding for named enums. I encoded it using Elixir's map, but the keyword list is also idiomatic.

```rust
#[rustler::NifEnum]
enum Test {
  foo{ x: i32, y: i32 },
}

#[rustler::nif]
fn claim() -> Test {
  Test::foo{ x: 1, y: 2 }
}
``` 

```elixir
# Current PR
iex(1)> claim()
%{__enum__: :foo, x: 1, y: 2}
# Use keyword list
iex(2)> claim()
{:foo, x: 1, y: 2}
```

**Comment by philss** [comment]
This feature looks pretty cool! I think it will be useful to me in a task I'm doing.

> I also want to have some conversation about encoding for named enums. I encoded it using Elixir's map, but the keyword list or a tuple contains the map are also idiomatic.

In my mind the version with `__enum__` looks a little bit weird. I would suggest one of the other two options you suggested. `Keyword` (2nd) seems slightly better because you can keep the order. But I'm OK with a map (3rd) too :)

**Comment by SeokminHong** [comment]
@philss Thanks for your reply!

After reading your comment, I think it will be more promising the 3rd option rather than the 2nd one. It's because, in Rust, the order of named enum does not matter.
So, in my mind, even though the `Keyword` (2nd) looks more simple, map syntax (3rd) makes more sense.

**Comment by evnu** [comment]
> It's because, in Rust, the order of named enum does not matter.

Also, the form resembles `NifMap` for structs. So if we use maps, the form is essentially the same and the enum variant is simply a tagged `NifMap`.

**Comment by SeokminHong** [comment]
I've implemented the third option and added more complex test cases!

**Comment by evnu** [comment]
@philss @SeokminHong  Wrt to unit variants, what do you prefer:

```
3 => Test::Qux,                      // {:qux}         OPTION 1
3 => Test::Qux,                      // :qux           OPTION 2
3 => Test::Qux,                      // {:qux, nil}   OPTION 3
```

Option 2 looks nicer, but Option 1 might be closer to the non-unit variant cases. Option 3 is closest to the other forms, `nil` would indicate that this does not carry any more data. Alas, Option 3 conflicts with a variant `Qux(Option<T>)`. Maybe there are better options?

**Comment by philss** [comment]
@evnu  I prefer the second option, with only the atom. The first one is strange, with a tuple of one element. But I would be OK with that too. I'm not sure how could be a forth option :thinking:

File: rustler_tests/test/codegen_test.exs
Comment (philss) [inline rustler_tests/test/codegen_test.exs:153]: This is so flexible that it seems wrong :sweat_smile: 
Loved it! :heart_eyes:

File: rustler_codegen/src/tagged_enum.rs
Comment (philss) [inline rustler_codegen/src/tagged_enum.rs:28]: Shouldn't we have an error message for the `expect("")` here?

Maybe:
```suggestion
                        let atom_str = field.ident.as_ref().expect("Atom is expected").to_string().to_snake_case();
```

**Review by philss — APPROVED** [review]
It looks great! Thank you, @SeokminHong!

I think we can deprecated `NifUnitEnum` and `NifUntaggedEnum` in another PR.

Let's wait for @evnu's review :)

**Comment by SeokminHong** [comment]
@evnu I also prefer the second option. It will also provide convenience for moving to `NifEnum` from `NifUnitEnum`.

**Comment by SeokminHong** [comment]
@evnu Before approving this review, could you review #448 first? If the PR merged, I would fix the `Result` type in this PR too.

**Comment by evnu** [comment]
>  It will also provide convenience for moving to NifEnum from NifUnitEnum.

Right, that's a good reason as well! I like it.

File: rustler_tests/test/codegen_test.exs
Comment (evnu) [inline rustler_tests/test/codegen_test.exs:145]: Please also add tests for failing decode. Each sub-type should be tested for error behaviour, e.g. `{:foo, "not a map"}`, `{:bar %{a: :map}}`, `{:baz, 10}`, `{:qux, :not_even_a_variant}`, `{:foo, :too, :many, :elements}`.

**Review by evnu — CHANGES_REQUESTED** [review]
A first review. I will need some more time for the internals of the encoder/decoder, but these comments might already provide some things which we need to do. All in all, this looks very promising! I wonder about some error cases (e.g. wrong datatypes for tagged data), but we should understand them better by extending the tests.

File: rustler_codegen/src/tagged_enum.rs
Comment (SeokminHong) [reply rustler_codegen/src/tagged_enum.rs:48]: I agree deduplicating is not needed. At first, I intended to prevent making the same atoms, but I think it makes codes more complicated and not actually necessary. 😃

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:48]: I think we don't need to test it if it already works.

File: rustler_tests/native/rustler_test/src/test_codegen.rs
Comment (SeokminHong) [reply rustler_tests/native/rustler_test/src/test_codegen.rs:93]: Thank you for catching it! I totally forgot to handle the unnamed enum fields having two or more. I added a transcoding example for `Bar2(i64, i64)`.

```elixir
  assert {:bar2, 1, 2} == RustlerTest.tagged_enum_2_echo({:bar2, 1, 2})
```

File: rustler_tests/test/codegen_test.exs
Comment (SeokminHong) [reply rustler_tests/test/codegen_test.exs:145]: What error type would be suited for these cases? `ErlangError` or `ArgumentError`?

Comment (evnu) [reply rustler_tests/test/codegen_test.exs:145]: Error handling should probably be quite similar to what `NifUntaggedEnum` returns if no variant matches

**Comment by SeokminHong** [comment]
@evnu, I fixed all of what you reviewed!

**Comment by evnu** [comment]
@SeokminHong thanks a lot for the changes! I am quite swamped right now, but will try to review soon!

**Comment by evnu** [comment]
@SeokminHong I really like where this is going! :) The `gen_decoder()` and `gen_encoder()` functions seem quite long, I think it might be beneficial to introduce helper functions and reduce the amount of code in those functions. I think that will also help with reviewing. What do you think?

**Review by evnu — CHANGES_REQUESTED** [review]
My comments are mostly nitpickings by now I think. Wrt error handling, specifying which field actually does not match after descending into a variant might be more user-friendly than a generic `invalid_variant` message.

File: rustler_codegen/src/tagged_enum.rs
Comment (SeokminHong) [reply rustler_codegen/src/tagged_enum.rs:109]: The Rust fmt automatically inserts them. Maybe `rustfmt.toml` is needed.

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:109]: Interesting, I would have expected `rustfmt` to remove them. Alright, just leave it as is! :)

File: rustler_codegen/src/tagged_enum.rs
Comment (SeokminHong) [reply rustler_codegen/src/tagged_enum.rs:307]: Then, how about changing the signature of `Term::map_from_arrays()` like this?

```diff
 pub fn map_from_arrays(
         env: Env<'a>,
-        keys: &[Term],
-        values: &[Term],
+        keys: &[impl Encoder],
+        values: &[impl Encoder],
     ) -> NifResult<Term<'a>>
```

In the previous signature, `.encode(env)` is needed for each elements.

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:307]: You are right, we would need to put in more work for that. As we use the same approach (`map_put()`) in `ex_struct.rs` as well, I think we should leave this improvement for a later PR.

Comment (SeokminHong) [reply rustler_codegen/src/tagged_enum.rs:307]: Ok, when the PR is merged, I'll make a new PR for improving functions arguments :)

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:307]: Thanks! I opened https://github.com/rusterlium/rustler/issues/451 to not forget about this :)

**Comment by SeokminHong** [comment]
@evnu I handled more detailed error messages for each decoding logic. However, there are some points that could be improved.

In the following test case, the enum expects the *payload* type is a map. But in this case, it passes a struct.

https://github.com/SeokminHong/rustler/blob/998e61a18547ab7263fae8a827595a3b5fec103b/rustler_tests/test/codegen_test.exs#L260-L263

The error message is `The map must have 2 elements, but it has 3` because the struct has a hidden field, `__struct__`. However, I think it's not intuitive. Do I improve this message to check the map has `__struct__` field and raise the proper message for this case?

**Comment by evnu** [comment]
>  Do I improve this message to check the map has __struct__ field and raise the proper message for this case?

The behaviour here should be similar to `NifMap`: If the map which is to be decoded contains all fields of the struct and the types match, decoding should work, even if the map contains more fields than actually required. So in this case, maybe the error should indeed tell the user, that a field is missing (simply the first field that we check). The exact number of fields is not that relevant I think.

**Comment by hansihe** [comment]
This all looks very very exciting! This has the potential to make these proc macros more powerful and easier to use.

Has there been any discussion about allowing attributes on the enum variants in order to specify its behavior?

One case I can imagine is wanting a struct variant in Rust to correspond to a tuple with an atom tag in Elixir. It might also allow you to specify the atom tag.

I have not thought about this very much (names and everything is just invented), but here is a quick example of what I mean:

```rust
#[rustler::NifEnum]
enum Test {
  // This might be the default?
  #[rustler(encoding = map_in_tuple)
  SomeVariant1 { x: i32, y: i32 },
  #[rustler(tag = "overridden_tag", encoding = tuple)
  SomeVariant2 { x: i32, y: i32 },
}
```

```
Test::SomeVariant1 { x: 1, y: 2 } => {:some_variant_1, %{x: 1, y: 2}}
Test::SomeVariant2 { x: 1, y: 2} => {:overridden_tag, 1, 2}
```

We might also want to allow specifying stuff on the fields within the enum variants.

We could potentially rework the other proc macros to unify (for example) `NifStruct` and `NifTuple` as well.

**Comment by evnu** [comment]
> Has there been any discussion about allowing attributes on the enum variants in order to specify its behavior?

@turion suggested using attributes to unify with structs earlier. Your suggestions look appealing and resemble serde, which is nice! I would probably move this discussion into a follow-up ticket in order to have this PR merged as a first working implementation towards something more unified. @hansihe What do you think? See also https://github.com/rusterlium/rustler/issues/426, which also aims to unify the different approaches.

> We could potentially rework the other proc macros to unify (for example) NifStruct and NifTuple as well.

Yes, we have a lot of overlap w.r.t. internals in the different macros.

**Comment by hansihe** [comment]
Yep, definitely agree it makes sense to get this merged and then work on expanding on it

File: rustler_codegen/src/tagged_enum.rs
Comment (evnu) [inline rustler_codegen/src/tagged_enum.rs:213]: I think we don't need that check, do we? If we cannot decode the field on index `i`, the decoder further above will report an error.

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:213]: (Or does this somehow behave differently then the named case?)

Comment (evnu) [reply rustler_codegen/src/tagged_enum.rs:213]: Scratch that, the error message probably already indicates to the user quite well what the problem is!

**Review by evnu — APPROVED** [review]

**Comment by evnu** [comment]
@SeokminHong thanks for all the work you put into this PR! Is there anything missing you wanted to add before we merge?

**Comment by SeokminHong** [comment]
I think it's fine! Thanks for all your feedback 😄

**Pull request merged**
